crates/bevy_graph/src/algos/dfs.rs

```diff
@@ -0,0 +1,132 @@
+use hashbrown::HashSet;
+
+use crate::graphs::{keys::NodeIdx, Graph};
+
+/// Implementation of the `DFS` algorythm
+///
+/// it will evaluate every node from the start as deep as it can and then up till the next node.
+pub struct DepthFirstSearch {
+    stack: Vec<NodeIdx>,
+    visited: HashSet<NodeIdx>,
+}
+
+impl DepthFirstSearch {
```

Comment:
It feels like these search methods should be refactored into a common trait, given the repetition and variety of potential search strategies available? I'll defer to @maniwani here though.

Reply:
I also thought about making them implement `Iterator`, but then i had to pass the graph at creation and wouldnt have unsfae methods. custom trait would be smart iguess

Reply:
> Yep, that was my first instinct too 🤔 You could have a `GraphIterator<G, S>` type though, and then store a graph of type `G` plus a search strategy of type `S` and implement `Iterator` for that?
> 
> Could be nice because then you get all the iterator convenience methods. And you could impl the unsafe methods directly on that type.

maybe, but i would keep this for an other pr